File: unity/Aetherlight.Core/Core/Registry.cs
using System.Collections.Generic;
using Aetherlight.Domain;

namespace Aetherlight.Core
{
    /// <summary>
    /// All game content is data, loaded from one or more packs. Packs merge in
    /// load order: a later pack overrides an earlier entry with the same id,
    /// which is what makes a balance patch or a mod pack possible without
    /// editing the base content.
    /// </summary>
    public sealed class ContentPack
    {
        public string Id = "";
        public string? Version;
        public List<ActorDef> Actors = new List<ActorDef>();
        public List<MoteDef> Motes = new List<MoteDef>();
        public List<ClassDef> Classes = new List<ClassDef>();
        public List<ArtDef> Arts = new List<ArtDef>();
        public List<GearDef> Gear = new List<GearDef>();
        public List<SummonDef> Summons = new List<SummonDef>();
        public List<StatusDef> Statuses = new List<StatusDef>();
        public List<EnemyDef> Enemies = new List<EnemyDef>();
    }

    public sealed class ValidationIssue
    {
        public bool IsError;
        public string Entity = "";
        public string Id = "";
        public string Message = "";

        public override string ToString() => $"{(IsError ? "error" : "warning")} {Entity} {Id}: {Message}";
    }

    public sealed class ContentRegistry : IActorContext
    {
        public readonly Dictionary<string, ActorDef> Actors = new Dictionary<string, ActorDef>();
        public readonly Dictionary<string, MoteDef> Motes = new Dictionary<string, MoteDef>();
        public readonly Dictionary<string, ClassDef> Classes = new Dictionary<string, ClassDef>();
        public readonly Dictionary<string, ArtDef> Arts = new Dictionary<string, ArtDef>();
        public readonly Dictionary<string, GearDef> Gear = new Dictionary<string, GearDef>();
        public readonly Dictionary<string, SummonDef> Summons = new Dictionary<string, SummonDef>();
        public readonly Dictionary<string, StatusDef> Statuses = new Dictionary<string, StatusDef>();
        public readonly Dictionary<string, EnemyDef> Enemies = new Dictionary<string, EnemyDef>();

        private readonly List<string> _packs = new List<string>();
        private List<ClassDef> _classList = new List<ClassDef>();

        public IReadOnlyList<string> Packs => _packs;

        public static ContentRegistry From(params ContentPack[] packs)
        {
            var registry = new ContentRegistry();
            foreach (var pack in packs) registry.Load(pack);
            return registry;
        }

        public ContentRegistry Load(ContentPack pack)
        {
            foreach (var entry in pack.Actors) Actors[entry.Id] = entry;
            foreach (var entry in pack.Motes) Motes[entry.Id] = entry;
            foreach (var entry in pack.Classes) Classes[entry.Id] = entry;
            foreach (var entry in pack.Arts) Arts[entry.Id] = entry;
            foreach (var entry in pack.Gear) Gear[entry.Id] = entry;
            foreach (var entry in pack.Summons) Summons[entry.Id] = entry;
            foreach (var entry in pack.Statuses) Statuses[entry.Id] = entry;
            foreach (var entry in pack.Enemies) Enemies[entry.Id] = entry;
            _packs.Add(pack.Id);
            _classList = new List<ClassDef>(Classes.Values);
            return this;
        }

        public ActorDef? ActorDef(string id) => Actors.TryGetValue(id, out var v) ? v : null;
        public MoteDef? MoteDef(string id) => Motes.TryGetValue(id, out var v) ? v : null;
        public ArtDef? ArtDef(string id) => Arts.TryGetValue(id, out var v) ? v : null;
        public GearDef? GearDef(string id) => Gear.TryGetValue(id, out var v) ? v : null;
        public SummonDef? SummonDef(string id) => Summons.TryGetValue(id, out var v) ? v : null;
        public StatusDef? StatusDef(string id) => Statuses.TryGetValue(id, out var v) ? v : null;
        public EnemyDef? EnemyDef(string id) => Enemies.TryGetValue(id, out var v) ? v : null;

        public IReadOnlyList<ClassDef> ClassDefs => _classList;

        /// <summary>
        /// Check every cross-reference. Run this in a content test so a typo in
        /// an art id fails the build instead of a boss fight.
        /// </summary>
        public List<ValidationIssue> Validate()
        {
            var issues = new List<ValidationIssue>();

            void RequireArt(string? id, string entity, string ownerId)
            {
                if (!string.IsNullOrEmpty(id) && !Arts.ContainsKey(id!))
                    issues.Add(new ValidationIssue { IsError = true, Entity = entity, Id = ownerId, Message = $"unknown art \"{id}\"" });
            }
            void RequireStatus(string? id, string entity, string ownerId)
            {
                if (!string.IsNullOrEmpty(id) && !Statuses.ContainsKey(id!))
                    issues.Add(new ValidationIssue { IsError = true, Entity = entity, Id = ownerId, Message = $"unknown status \"{id}\"" });
            }

            foreach (var actor in Actors.Values)
                foreach (var artId in actor.BaseArts) RequireArt(artId, "actor", actor.Id);

            foreach (var cls in Classes.Values)
                foreach (var grant in cls.Arts) RequireArt(grant.ArtId, "class", cls.Id);

            foreach (var art in Arts.Values)
                foreach (var status in art.Effect.Statuses) RequireStatus(status.StatusId, "art", art.Id);

            foreach (var gear in Gear.Values) RequireArt(gear.Proc?.ArtId, "gear", gear.Id);

            foreach (var enemy in Enemies.Values)
            {
                foreach (var artId in enemy.Arts) RequireArt(artId, "enemy", enemy.Id);
                foreach (var drop in enemy.Drops)
                {
                    if (!Gear.ContainsKey(drop.ItemId))
                        issues.Add(new ValidationIssue { IsError = false, Entity = "enemy", Id = enemy.Id, Message = $"drop \"{drop.ItemId}\" is not gear" });
                }
            }

            foreach (var summon in Summons.Values)
            {
                if (Domain.Summons.CostTotal(summon) == 0)
                    issues.Add(new ValidationIssue { IsError = false, Entity = "summon", Id = summon.Id, Message = "summon costs nothing" });
            }

            return issues;
        }

        public void AssertValid()
        {
            var errors = Validate().FindAll(i => i.IsError);
            if (errors.Count == 0) return;
            var detail = string.Join("\n", errors.ConvertAll(e => "  " + e));
            throw new System.InvalidOperationException($"Content validation failed with {errors.Count} error(s):\n{detail}");
        }
    }
}

File: src/core/registry.ts
/**
 * The content registry.
 *
 * All game content - actors, motes, classes, arts, gear, summons, statuses,
 * enemies - is data, loaded from one or more content packs. The rules layer
 * only ever reads through this registry, so a project can ship its own packs,
 * merge community packs on top, or hot-reload during development without the
 * systems knowing anything changed.
 *
 * Packs merge in load order: a later pack overrides an earlier entry with the
 * same id. That is what makes a "balance patch" pack or a mod pack possible
 * without editing the base content.
 */

import type { ActorDef } from '../domain/actor.js';
import type { ArtDef } from '../domain/arts.js';
import type { ClassDef } from '../domain/classes.js';
import type { GearDef } from '../domain/equipment.js';
import type { MoteDef } from '../domain/motes.js';
import type { StatusDef } from '../domain/status.js';
import type { SummonDef } from '../domain/summons.js';
import type { EnemyDef } from '../domain/enemy.js';
import type { ItemDef } from '../domain/items.js';

export interface ContentPack {
  id: string;
  version?: string;
  actors?: ActorDef[];
  motes?: MoteDef[];
  classes?: ClassDef[];
  arts?: ArtDef[];
  gear?: GearDef[];
  summons?: SummonDef[];
  statuses?: StatusDef[];
  enemies?: EnemyDef[];
  items?: ItemDef[];
}

export interface ValidationIssue {
  severity: 'error' | 'warning';
  packId: string;
  entity: string;
  id: string;
  message: string;
}

export class ContentRegistry {
  readonly actors = new Map<string, ActorDef>();
  readonly motes = new Map<string, MoteDef>();
  readonly classes = new Map<string, ClassDef>();
  readonly arts = new Map<string, ArtDef>();
  readonly gear = new Map<string, GearDef>();
  readonly summons = new Map<string, SummonDef>();
  readonly statuses = new Map<string, StatusDef>();
  readonly enemies = new Map<string, EnemyDef>();
  readonly items = new Map<string, ItemDef>();

  private readonly loadedPacks: string[] = [];

  static from(...packs: ContentPack[]): ContentRegistry {
    const registry = new ContentRegistry();
    for (const pack of packs) registry.load(pack);
    return registry;
  }

  load(pack: ContentPack): this {
    put(this.actors, pack.actors);
    put(this.motes, pack.motes);
    put(this.classes, pack.classes);
    put(this.arts, pack.arts);
    put(this.gear, pack.gear);
    put(this.summons, pack.summons);
    put(this.statuses, pack.statuses);
    put(this.enemies, pack.enemies);
    put(this.items, pack.items);
    this.loadedPacks.push(pack.id);
    return this;
  }

  get packs(): readonly string[] {
    return this.loadedPacks;
  }

  /** Class list as an array, which is what class resolution wants. */
  get classList(): ClassDef[] {
    return [...this.classes.values()];
  }

  actorDef = (id: string): ActorDef | undefined => this.actors.get(id);
  moteDef = (id: string): MoteDef | undefined => this.motes.get(id);
  artDef = (id: string): ArtDef | undefined => this.arts.get(id);
  gearDef = (id: string): GearDef | undefined => this.gear.get(id);
  summonDef = (id: string): SummonDef | undefined => this.summons.get(id);
  statusDef = (id: string): StatusDef | undefined => this.statuses.get(id);
  enemyDef = (id: string): EnemyDef | undefined => this.enemies.get(id);
  itemDef = (id: string): ItemDef | undefined => this.items.get(id);

  /** Context object accepted by the actor derivation functions. */
  actorContext() {
    return {
      actorDef: this.actorDef,
      moteDef: this.moteDef,
      classDefs: this.classList,
      gearDef: this.gearDef,
      statusDef: this.statusDef,
    };
  }

  /**
   * Check every cross-reference. Run this in a content test so a typo in an
   * art id fails the build instead of throwing halfway through a boss fight.
   */
  validate(): ValidationIssue[] {
    const issues: ValidationIssue[] = [];
    const packId = this.loadedPacks.join('+') || '(none)';

    const requireArt = (id: string | undefined, entity: string, ownerId: string) => {
      if (id && !this.arts.has(id)) {
        issues.push({ severity: 'error', packId, entity, id: ownerId, message: `unknown art "${id}"` });
      }
    };
    const requireStatus = (id: string | undefined, entity: string, ownerId: string) => {
      if (id && !this.statuses.has(id)) {
        issues.push({ severity: 'error', packId, entity, id: ownerId, message: `unknown status "${id}"` });
      }
    };

    for (const actor of this.actors.values()) {
      for (const artId of actor.baseArts ?? []) requireArt(artId, 'actor', actor.id);
    }
    for (const cls of this.classes.values()) {
      for (const grant of cls.arts ?? []) requireArt(grant.artId, 'class', cls.id);
    }
    for (const art of this.arts.values()) {
      for (const status of art.effect.statuses ?? []) requireStatus(status.statusId, 'art', art.id);
    }
    for (const mote of this.motes.values()) {
      if (mote.effect.kind === 'status') requireStatus(mote.effect.statusId, 'mote', mote.id);
    }
    for (const gear of this.gear.values()) {
      requireArt(gear.proc?.artId, 'gear', gear.id);
    }
    for (const enemy of this.enemies.values()) {
      for (const artId of enemy.arts ?? []) requireArt(artId, 'enemy', enemy.id);
    }
    for (const item of this.items.values()) {
      requireArt(item.artId, 'item', item.id);
      if (item.kind === 'consumable' && !item.artId) {
        issues.push({ severity: 'warning', packId, entity: 'item', id: item.id, message: 'consumable does nothing when used' });
      }
    }
    for (const enemy of this.enemies.values()) {
      for (const drop of enemy.drops ?? []) {
        // A drop naming a real item is the common case now that items exist;
        // the gear fallback stays for packs written before they did.
        if (!this.items.has(drop.itemId) && !this.gear.has(drop.itemId)) {
          issues.push({ severity: 'warning', packId, entity: 'enemy', id: enemy.id, message: `drop "${drop.itemId}" is neither an item nor gear` });
        }
      }
    }
    for (const summon of this.summons.values()) {
      if (Object.values(summon.cost).every((n) => !n)) {
        issues.push({ severity: 'warning', packId, entity: 'summon', id: summon.id, message: 'summon costs nothing' });
      }
    }
    return issues;
  }

  /** Throws if validation found any error-severity issue. */
  assertValid(): void {
    const errors = this.validate().filter((issue) => issue.severity === 'error');
    if (errors.length > 0) {
      const detail = errors.map((e) => `  ${e.entity} ${e.id}: ${e.message}`).join('\n');
      throw new Error(`Content validation failed with ${errors.length} error(s):\n${detail}`);
    }
  }
}

function put<T extends { id: string }>(target: Map<string, T>, entries: T[] | undefined): void {
  for (const entry of entries ?? []) target.set(entry.id, entry);
}
